### PR TITLE
Add token field for exact app matches

### PR DESCRIPTION
**Pull request by jameswestman**
Adds an `apps` field to repo tokens. `has_token_prefix` checks it in addition to the `prefixes` field, but only for exact matches. This will allow us to, for example, give access to an app but not its extensions.

### CODE REVIEW

**Comment by jameswestman** [comment]
cc @danielsilverstone-ct

File: src/api.rs [hunk 1]
```diff
@@ -135,6 +136,13 @@ pub fn prefix_is_subset(
     }
 }
 
+pub fn apps_is_subset(maybe_subset_apps: &Option<Vec<String>>, claimed_apps: &[String]) -> bool {
```
Comment (danielsilverstone-ct) [inline src/api.rs:139]: My gut feeling is that it'd be slightly nicer if this were:
```suggestion
pub fn apps_is_subset(maybe_subset_apps: Option<&[String]>, claimed_apps: &[String]) -> bool {
```

File: src/api.rs [hunk 2]
```diff
@@ -148,6 +156,7 @@ pub fn token_subset(
             && tokens::sub_has_prefix(&args.sub, &claims.sub)
             && args.scope.iter().all(|s| claims.scope.contains(s))
             && prefix_is_subset(&args.prefixes, &claims.prefixes)
+            && apps_is_subset(&args.apps, &claims.apps)
```
Comment (danielsilverstone-ct) [inline src/api.rs:159]: To achieve the above suggestion, you'd need something like:

```suggestion
            && apps_is_subset(args.apps.as_deref(), &claims.apps)
```

File: src/api.rs [hunk 3]
```diff
@@ -161,6 +170,13 @@ pub fn token_subset(
                         claims.prefixes.clone()
                     }
                 },
+                apps: {
+                    if let Some(ref apps) = args.apps {
+                        apps.clone()
+                    } else {
+                        claims.apps.clone()
+                    }
+                },
```
Comment (danielsilverstone-ct) [inline src/api.rs:179]: While this pattern is the same as others in this block, I can't help but feel that something along the lines of:

```suggestion
                apps: args.apps.clone().unwrap_or_else(|| claims.apps.clone()),
```

might work more nicely.  Of course, I'm more used to options etc. and if you feel the expanded out approach is clearer for newcomers to the code, I'm happy to be ignored on this one.

**Review by danielsilverstone-ct — APPROVED** [review]
Codewise I think this is fine - I made a few code suggestions, but I'm comfortable if none of them are taken.

File: src/api.rs (hunk 3 above)
Comment (jameswestman) [reply src/api.rs:179]: I'll leave it the same as the others for now, but maybe change it in the future.

**Review by danielsilverstone-ct — APPROVED** [review]

**Pull request merged**